### ISSUE TITLE
gradient layout contract

### ISSUE THREAD

**Issue by credoxfromcqut** [pytorch/pytorch#57706]
## 🐛 Bug

# **Warning: grad and param do not obey the gradient layout contract. This is not an error, but may impair performance.**

when the warnning appear after a valid function, the gpu occupation 70% goes to 100%.and cpu ocupation 50% goes to 10%,
the valid function just do a forward pass and it will change the model's mode(train or eval) and recover the original mode.then ,after the loss computation done, loss.backward() excute ,the warnning shows.besides, the loss computation costs much time after this
```
dist_ap, relative_p_inds = torch.max(
    dist_mat[is_pos].contiguous().view(N, -1), 1, keepdim=True)  # costly
```
when annotate the valid function， no warnning, the  cpu and gpu  occupation keeps almost constant.
after my step-wise test on the code above, **dist_mat[is_pos]** takes much time,the key is that the size of it is fixed after and before the valid function...,so i have to ensure this dont bring error result.
## To Reproduce
[here is the train source](https://github.com/huanghoujing/person-reid-triplet-loss-baseline/blob/a1c2bd20180cfaf225782717aeaaab461798abec/script/experiment/train.py)

## Environment
 - PyTorch Version (e.g., 1.0):1.7.0
 - OS (e.g., Linux):windows
 - How you installed PyTorch (`conda`, `pip`, source):--
 - Build command you used (if compiling from source):
 - Python version:3.6
 - CUDA/cuDNN version:
 - GPU models and configuration:1660super
 - Any other relevant information:

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano

**Comment by credoxfromcqut** [comment]
> what is the gradient layout contract?

sorry i dont know,so i even dont know the direction to think on!!

**Comment by albanD** [comment]
As mentioned here, this is not a correctness problem but just perf as you observed.
You can try to use `optimizer.zero_grad(set_to_none=True)` to fully reset the gradients at each iteration. That might help if you are actually change the layout of the weights/grads without knowing it.
